velox/functions/sparksql/String.h

```diff
@@ -455,4 +455,95 @@ struct SubstrFunction {
   }
 };
 
+template <typename T>
+struct OverlayFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    doCall<false>(result, input, replace, pos, len);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    doCall<true>(result, input, replace, pos, len);
+  }
+
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      const arg_type<Varchar>& replace,
+      const int32_t pos,
+      const int32_t len) {
+    // Calculate and append first part.
+    auto pair = substring<isAscii>(input, 1, pos - 1);
+    auto byteRange = stringCore::getByteRange<isAscii>(
+        input.data(), pair.first + 1, pair.second);
+    result.append(StringView(
+        input.data() + byteRange.first, byteRange.second - byteRange.first));
+
+    // Append second part.
+    result.append(replace);
+
+    // Calculate and append last part.
+    int32_t length = len >= 0 ? len : stringImpl::length<isAscii>(replace);
+    int64_t start = (int64_t)pos + (int64_t)length;
+    pair = substring<isAscii>(input, start, INT32_MAX);
+    byteRange = stringCore::getByteRange<isAscii>(
+        input.data(), pair.first + 1, pair.second);
+    result.append(StringView(
+        input.data() + byteRange.first, byteRange.second - byteRange.first));
+  }
+
+  // Information regarding the pos calculation:
+  // Hive and SQL use one-based indexing for SUBSTR arguments but also accept
+  // zero and negative indices for start positions. If a start index i is
+  // greater than 0, it refers to element i-1 in the sequence. If a start index
+  // i is less than 0, it refers to the -ith element before the end of the
```

Comment:
This doesn't match the documentation which says that negative pos means start of the string.

Reply:
This method and comments were port from Spark, argument `pos` here is different with overlay function's argument.